python/analyzers/HNLAnalyzer.py

```diff
@@ -0,0 +1,174 @@
+            mm.pdgId  = lambda : -(mm.charge()*13)
+        
+        for jj, mm in enumerate(event.dSAMu):
+            mm.track = lambda : ROOT.reco.TrackRef(self.handles['dSAMu'].product(),jj)
+            
+        
+        # store the number of sMu and dSAMu per event
+        if len(event.sMu):
+            event.n_sMu = len(event.sMu)
+            # event.maxptsMu=sorted(event.sMu,key=lambda x: x.pt(),reverse = True)[0] 
+        else:
+            event.n_sMu = 0    
+
+        if len(event.dSAMu):
+            event.n_dSAMu = len(event.dSAMu)
+            # event.maxptdSAMu=sorted(event.dSAMu,key=lambda x: x.pt(),reverse=True)[0]
+        else:
+            event.n_dSAMu = 0
+        
+        ##############
+        # for now ONLY work with sMu
+        ##############
+
+        # select only events with >= 3 muons
+        if event.n_sMu < 3:
+            return False
+
+        self.counters.counter('HNL').inc('>= 3 muons')
+       
+        # select only events with OS muon pairs and collect the pairs
+        event.os_pairs = [pair for pair in combinations(event.sMu,2) if pair[0].charge() != pair[1].charge()] 
+
+        if not len(event.os_pairs):
+            return False
+
+        self.counters.counter('HNL').inc('>= 3 muons && > 0 os_pairs')
+
+        # only selecting dimuon pairs with mutual vertices (surviving the kinematic vertex fitter) 
+        dimuon = []
+        for pair in event.os_pairs:
+            self.tofit.clear()
+            for il in pair:
+                myp4 = ROOT.Math.LorentzVector('<ROOT::Math::PxPyPzE4D<double> >')(il.px(), il.py(), il.pz(), sqrt(il.mass()**2 + il.px()**2 + il.py()**2 + il.pz()**2))
+                ic = ROOT.reco.RecoChargedCandidate() # instantiate a dummy RecoChargedCandidate
+                ic.setCharge(il.charge())           # assign the correct charge
+                ic.setP4(myp4)                      # assign the correct p4
+                ic.setTrack(il.track())             # set the correct TrackRef
+                if ic.track().isNonnull():          # check that the track is valid, there are photons around too!
+                    self.tofit.push_back(ic)
+            # further sanity check: two *distinct* tracks
+            if self.tofit.size() == 2 and self.tofit[0].track() != self.tofit[1].track:
+                svtree = self.vtxfit.Fit(self.tofit) # the actual vertex fitting!
+                if not svtree.get().isEmpty() and svtree.get().isValid(): # check that the vertex is good
+                    svtree.movePointerToTheTop()
+                    sv = svtree.currentDecayVertex().get()
+                    # if sv.event.recoSv.Chi2() 
+                    # event.recoSv = makeRecoVertex(sv, kinVtxTrkSize=2)
+                    dimuon.append(DiMuon(pair, makeRecoVertex(sv, kinVtxTrkSize=2)))
+
+        if not len(dimuon):
+            return False
+        
+        self.counters.counter('HNL').inc('>= 3 muons && > 0 os_pairs with mutual vertices')
```

Comment:
to keep the counter cleaner and easier to read, I suggest to drop the N-1 cuts.
In practice, instead of 
```
self.counters.counter('HNL').inc('>= 3 muons && > 0 os_pairs with mutual vertices')
```
just do
```
self.counters.counter('HNL').inc('os_pairs with mutual vertices')
```